### PR TITLE
#31-implement-baserepository

### PR DESCRIPTION
**Pull request by NodirbekAzizov-ACC**
Base Repository is implemented in shared solution, Interface of Base Repository is in the same folder with its implementation.
Mongo Db settings is configured.
Ready to review.

### CODE REVIEW

File: src/BudgetManager/BudgetManager.Model/Transaction.cs [hunk 1]
```diff
@@ -7,7 +7,6 @@
 
 namespace BudgetManager.Model
 {
-    [CollectionName("Transactions")]
```
Comment (Saigunas) [inline src/BudgetManager/BudgetManager.Model/Transaction.cs]: Why did you remove the collection name?

Comment (NodirbekAzizov-ACC) [reply src/BudgetManager/BudgetManager.Model/Transaction.cs]: Sorry, by mistake, i will fix that

Comment (NodirbekAzizov-ACC) [reply src/BudgetManager/BudgetManager.Model/Transaction.cs]: let me first collect  all required changes

**Comment by Saigunas** [comment]
Also I added a hashtag to the PR title.

File: src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+namespace BudgetManager.Shared.DataAccess.MongoDB.BaseImplementation
+{
+    public class BaseRepository<TDocument> : IBaseRepository<TDocument>
+        where TDocument : IModelBase
+    {
+        private readonly IMongoCollection<TDocument> _collection;
+        public BaseRepository(IMongoDbSettings settings)
+        {
+            var database = new MongoClient(settings.ConnectionString).GetDatabase(settings.DatabaseName);
+            _collection = database.GetCollection<TDocument>(GetCollectionName(typeof(TDocument)));
```
Comment (VladislavGromadtsov) [inline src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs:18]: Need to add following string here because our Id is storing like Binary not UUID
```suggestion
            BsonSerializer.RegisterSerializer(new GuidSerializer(GuidRepresentation.Standard));
            var database = new MongoClient(settings.ConnectionString).GetDatabase(settings.DatabaseName);
            _collection = database.GetCollection<TDocument>(GetCollectionName(typeof(TDocument)));
```

**Review by VladislavGromadtsov — CHANGES_REQUESTED** [review]

**Review by VladislavGromadtsov — APPROVED** [review]

File: src/BudgetManager/BudgetManager.Model/Transaction.cs (hunk 1 above)
Comment (Saigunas) [inline src/BudgetManager/BudgetManager.Model/Transaction.cs]: Don't forget this

**Review by Saigunas — CHANGES_REQUESTED** [review]

**Review by sirojbekdev — APPROVED** [review]

**Review by SokhibJones — APPROVED** [review]

File: src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs [hunk 3]
```diff
@@ -2,20 +2,18 @@
...
 using MongoDB.Bson;
-using MongoDB.Bson.Serialization;
-using MongoDB.Bson.Serialization.Serializers;
 using MongoDB.Driver;
 using MongoDbGenericRepository.Attributes;
 
 namespace BudgetManager.Shared.DataAccess.MongoDB.BaseImplementation
 {
-    public class BaseRepository<TDocument> : IBaseRepository<TDocument>
+    public abstract class BaseRepository<TDocument>
```
Comment (Saigunas) [inline src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs]: Why did you remove the interface and make the class abstract?

Comment (NodirbekAzizov-ACC) [reply src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs]: you can inherit it and use its methods, it case if they not full fill business requirements, then you can override any

Comment (Saigunas) [reply src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs]: Hm, there was no problem overriding methods before too though. I think we should let Doniyor review this PR and have his say. Idk if this is a good practice.

**Review by Saigunas — APPROVED** [review]

File: src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs (hunk 3 above)
Comment (VladislavGromadtsov) [reply src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs]: According to the SOLID principles
In case of abstract class we will broke Dependency Inversion principle. High-level modules should not depend on lower-level modules.  Both should depend on abstractions. Abstractions should not depend on details. Details should depend on abstractions. 

In this case our repos shouldn't depend on the lower-level class (even if it's an abstract class) to get its work done. Both the lower-level and higher-level class should depend on the interface which defines how they are going to interact.

Our BaseRep should have only common things to implement, if we need smth to add - we should add it in our rep which will be inherited from BaseRep

We should have loosely coupled entities that are easy to test, modify and update.

Comment (NodirbekAzizov-ACC) [reply src/BudgetManager.Shared/BudgetManager.Shared.DataAccess.MongoDB/BaseImplementation/BaseRepository.cs]: I see,
Ok, let me fix that

**Review by otash8110 — APPROVED** [review]
